### ISSUE TITLE
`complete -f` doesn't play well with `__fish_complete_path`

### ISSUE THREAD

**Issue by fvictorio** [fish-shell/fish-shell#7530]
- `fish --version`: `fish, version 3.1.0`
- `echo $version`: `3.1.0`
- `uname -a`: `Linux leary 5.4.0-54-generic #60-Ubuntu SMP Fri Nov 6 10:37:59 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux`
- `echo $TERM`: `xterm-kitty`
- My `fish` config is already minimal, it only has the steps necessary to reproduce this.

---

I'm writing a completion script that delegates the completion to the program that is being completed. This is done with this command (`hh` is the program being completed, and `_hh_completion` is the function that calls the program, gets the suggestions and prints them):

```fish
complete -f -d 'hh' -c hh -a "(eval _hh_completion)"
```

Inside the `_hh_completion`, I have a condition where I just want to tell fish to complete a path file, so I use `__fish_complete_path`. This works fine, **but only for files at the first level of the current working directory**. In other words, I can complete a directory name, but not a file inside that directory. Here's an asciinema illustrating the problem:

https://asciinema.org/a/wtG4qKIF28DVhxM5mADXwHiVq

Notice how `node_modules` is completed fine, but when I press `<tab>` after that, it inserts an space.

Here's the full completion script:

```fish
###-begin-hh-completion-###
function _hh_completion
  set cmd (commandline -o)
  set cursor (commandline -C)
  set words (node -pe "'$cmd'.split(' ').length")

  set completions (eval env DEBUG=\"" \"" COMP_CWORD=\""$words\"" COMP_LINE=\""$cmd \"" COMP_POINT=\""$cursor\"" hh-completion completion -- $cmd)

  if [ "$completions" = "__tabtab_complete_files__" ]
    __fish_complete_path
  else
    for completion in $completions
      echo -e $completion
    end
  end
end

complete -f -d 'hh' -c hh -a "(eval _hh_completion)"
###-end-hh-completion-###
```

---

A "simpler" way to reproduce this problem is doing this:

```
complete -f -d 'foo' -c foo -a "(eval __fish_complete_path)"
```

and then doing

```
foo some-dir/<tab>
```

**Comment by faho** [comment]
1. You have about 2 `eval`s too many here - you can just run `__fish_complete_path`, no need to `eval` it.

2. If you want `__fish_complete_path` to complete anything but the current path, you'll have to pass that path to it:

```fish
complete -f -d 'foo' -c foo -a "(__fish_complete_path (commandline -ct))"
```

3. `set words (node -pe "'$cmd'.split(' ').length")` is just `set words (count $cmd)`

**Comment by fvictorio** [comment]
This worked fantastically. Thanks a lot for the help @faho!

It seems to me that `__fish_complete_path (commandline -ct)` is what anyone using `__fish_complete_path` would want, so maybe it makes sense to add it somewhere in the docs.

Anyway, closing this issue, thanks again!

**Comment by fvictorio** [comment]
Hi, sorry for bothering you again, but I think there's an issue with the `__fish_complete_path (commandline -ct)` approach. If you start writing some text and press tab, but no file starts with that string, you get this error:

```
/usr/share/fish/functions/__fish_complete_path.fish (line 1): No matches for wildcard “"$target"*”. See `help expand`.
command ls -dp "$target"*
               ^
in command substitution
	called on line 13 of file /usr/share/fish/functions/__fish_complete_path.fish
in function '__fish_complete_path' with arguments 'fo'
in command substitution
/usr/share/fish/functions/__fish_complete_path.fish: Unknown error while evaluating command substitution
in function '__fish_complete_path' with arguments 'fo'
in command substitution
```

I can open a new issue for this, but I thought I'd ask here first just in case. Thank you!

**Comment by krobelus** [comment]
@fvictorio should be fixed with f1478137dc706a9fc16b73126e260b340e6d3c0b
